openmeter/ent/schema/subscription.go

```diff
@@ -0,0 +1,141 @@
+	"github.com/openmeterio/openmeter/pkg/datex"
+	"github.com/openmeterio/openmeter/pkg/framework/entutils"
+)
+
+type Subscription struct {
+	ent.Schema
+}
+
+func (Subscription) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.NamespaceMixin{},
+		entutils.TimeMixin{},
+		entutils.MetadataAnnotationsMixin{},
+		entutils.CadencedMixin{},
+	}
+}
+
+func (Subscription) Fields() []ent.Field {
+	return []ent.Field{
+		field.String("plan_key").NotEmpty().Immutable(),
+		field.Int("plan_version").Min(1).Immutable(),
+		field.String("customer_id").NotEmpty().Immutable(),
+		field.String("currency").GoType(currencyx.Code("")).MinLen(3).MaxLen(3).NotEmpty().Immutable(),
+	}
+}
+
+func (Subscription) Indexes() []ent.Index {
+	return []ent.Index{
+		index.Fields("namespace", "id"),
+		index.Fields("namespace", "customer_id"),
+	}
+}
+
+func (Subscription) Edges() []ent.Edge {
+	return []ent.Edge{
+		edge.To("entitlements", SubscriptionEntitlement.Type),
+		edge.From("customer", Customer.Type).Field("customer_id").Ref("subscription").Immutable().Unique().Required(),
+		edge.To("phases", SubscriptionPhase.Type),
+	}
+}
+
+type SubscriptionPhase struct {
+	ent.Schema
+}
+
+func (SubscriptionPhase) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.NamespaceMixin{},
+		entutils.TimeMixin{},
+		entutils.MetadataAnnotationsMixin{},
+	}
+}
+
+func (SubscriptionPhase) Fields() []ent.Field {
+	return []ent.Field{
+		field.String("subscription_id").NotEmpty().Immutable(),
+		field.String("key").NotEmpty().Immutable(),
+		field.String("name").NotEmpty(),
+		field.String("description").Optional().Nillable(),
+		field.Time("active_from").Immutable(),
+	}
+}
+
+func (SubscriptionPhase) Indexes() []ent.Index {
+	return []ent.Index{
+		index.Fields("namespace", "id"),
+		index.Fields("namespace", "subscription_id"),
+		index.Fields("namespace", "subscription_id", "key"),
+	}
+}
+
+func (SubscriptionPhase) Edges() []ent.Edge {
+	return []ent.Edge{
+		edge.From("subscription", Subscription.Type).Field("subscription_id").Ref("phases").Unique().Immutable().Required(),
+		edge.To("items", SubscriptionItem.Type),
+	}
+}
+
+type SubscriptionItem struct {
+	ent.Schema
+}
+
+func (SubscriptionItem) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.NamespaceMixin{},
+		entutils.TimeMixin{},
+		entutils.MetadataAnnotationsMixin{},
+	}
+}
+
+func (SubscriptionItem) Fields() []ent.Field {
+	return []ent.Field{
+		// Note: we cannot use CadencedMixin as these are mutable
+		field.Time("active_from"),
+		field.Time("active_to").Optional().Nillable(),
+		field.String("phase_id").NotEmpty().Immutable(),
+		field.String("start_after_phase").
+			GoType(datex.ISOString("")).Nillable().Optional(),
+		field.String("end_after_phase").
+			GoType(datex.ISOString("")).Nillable().Optional(),
+		field.String("key").NotEmpty().Immutable(),
+		field.String("rate_card").
+			GoType(&subscription.RateCard{}).
```

Comment:
any reason it's not materialized? In product catalog rate card is materialized:
https://github.com/openmeterio/openmeter/blob/main/openmeter/ent/schema/productcatalog.go#L119

Reply:
Right now mainly for simplicity and time saving. ProductCatalog models (like RateCard) should be made available for use outside of those packages which is currently not possible, the `subscription.RateCard` should disappear once that is resolved. I'd be much happier to reuse the exact code used in Plan for much of this, but you're right that's not a requirement, I'm happy to normalize it if you'd like, it's just a bit of double work, but there's already quite a bit of double work happening so that doesn't have to be an issue.